src/theme.c

```diff
@@ -950,6 +978,186 @@ create_corners(struct theme *theme)
+ * from `start_color` at the top-left to clear at the opposite edge. A cut-out
+ * is added in the top-left to stop the shadow appearing behind the window, the
+ * corner radius of this cutout is set by `cutout_corner_radius`.
+ */
+static void
+shadow_corner_gradient(cairo_t *cr, int radius, double inset,
+	int cutout_corner_radius, float start_color[4])
+{
+	float end_color[4] = {
+		start_color[0], start_color[1], start_color[2], 0.0};
+
+	/*
+	 * The inset passed is how big the inset is as a proportion of the
+	 * visible shadow width. But for the gradient we want to know how big
+	 * the inset is as a proportion of total shadow width.
+	 */
+	inset = inset / (1.0 + inset);
+
+	cairo_surface_t *surf = cairo_get_target(cr);
+	cairo_pattern_t *pat = cairo_pattern_create_radial(
+		0.0, 0.0, 0.0, 0.0, 0.0, (double)radius);
+
+	cairo_pattern_add_color_stop_rgba(pat, 0.0,
+		start_color[0], start_color[1], start_color[2], start_color[3]);
+	/*
+	 * The shadow should only start fading from the window edge, otherwise
+	 * the configured inset effectively reduces opacity.
+	 */
+	cairo_pattern_add_color_stop_rgba(pat, inset,
+		start_color[0], start_color[1], start_color[2], start_color[3]);
+	cairo_pattern_add_color_stop_rgba(pat, 1.0,
+		end_color[0], end_color[1], end_color[2], end_color[3]);
+
+	/* As in shadow_edge_gradient(), trick to approximate Gaussian blur */
+	cairo_save(cr);
+	cairo_rectangle(cr, 0.0, 0.0, (double)radius, (double)radius);
+	cairo_clip(cr);
+	cairo_set_source(cr, pat);
+	cairo_mask(cr, pat);
+	cairo_restore(cr);
+
+	/*
+	 * Finally, erase the portion of the shadow which will be inset behind
+	 * the window otherwise things will look odd of the window is
+	 * translucent! The inset is just to make the corner gradient look
+	 * right, it should never actually be seen.  The rectangle we erase
+	 * might need a rounded corner to match a rounded titlebar.
+	 */
+	double pi = 3.142;
+	cairo_save(cr);
+	cairo_set_operator(cr, CAIRO_OPERATOR_CLEAR);
+	/* top left */
+	cairo_move_to(cr, 0.0, 0.0);
+	/* top right */
+	cairo_line_to(cr, inset * radius, 0.0);
+	/* bottom right, with radius */
+	cairo_arc(cr, inset * radius - cutout_corner_radius,
+		inset * radius - cutout_corner_radius,
+		cutout_corner_radius, 0.0, pi / 2.0);
+	/* bottom left */
+	cairo_line_to(cr, 0.0, inset * radius);
+	cairo_close_path(cr);
+	cairo_fill(cr);
+	cairo_restore(cr);
+
+	cairo_pattern_destroy(pat);
+	cairo_surface_flush(surf);
+}
+
+static void
+create_shadows(struct theme *theme)
+{
+	int inset_active = (double)theme->window_active_shadow_radius
+		* theme->window_active_shadow_inset;
+	int inset_inactive = (double)theme->window_inactive_shadow_radius
+		* theme->window_inactive_shadow_inset;
+	/* Total width including visible and obscured portion */
+	int total_active_width =
+		theme->window_active_shadow_radius + inset_active;
+	int total_inactive_width =
+		theme->window_inactive_shadow_radius + inset_inactive;
```

Comment:
It seems perfectly reasonable to allow a user to set a zero radius and inset for one or both of these parameters and react accordingly. For example, one might enable shadows globally and completely disable them in the theme, or disable only inactive window shadows by setting the radius and inset to 0. When one of `total_active_width` or `total_inactive_width` is zero, I suggest you avoid trying to create all buffers associated with the corresponding shadow, and then adapt other logic to avoid trying to draw shadows for that case.

The parsing of the themerc should properly validate all numeric arguments. For example, a single negative value will crash labwc, while two negative values will efefctively disable shadows but allow the program to continue. When parsing these parameters, they should be clipped to zero and a warning issued if negative values are specified.

Reply:
We do have `post_processing()` for that and I agree that the values should be validated / clamped.

It is also a broader issue in labwc in general, we likely have quite some more ints that aren't being checked but that is unrelated this PR.